69-common-kingslayer/technical_specification.md

```diff
@@ -0,0 +1,51 @@
+# OpenTelemetry in file services (Common Kingslayer)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+## Scope
+### Outline:
+
+Based on some previous exploration work, this epic aims to implement 1) OpenTelemetry instrumentation in the file services and 2) necessary OpenTelemetry functionality inside of hexkit and/or ghga-service-commons.
+In addition, a viable setup of Jaeger and its components needs to be devised and implemented.
+
+### Included/Required:
+
+While OpenTelemetry supports tracing, metrics and logging, this epic will focus on enabling collection of distributed tracing data and neglect the other two aspects.
+
+#### File Services:
+File services can export tracing data using manual or autoinstrumentation or a combination of both.
+Autoinstrumentation works by installing specific libraries that monkeypatch existing functionality to inject information that can be propagated across service boundaries.
+While these are quite quite useful, they should be vetted for what data is attached to the traces and manual instrumentation should be used instead, if sensitive information is included.
+If autoinstrumentation can be used in all/most cases, additional implementation work can be kept to a minimum.
+
+#### Hexkit/Service Commons:
+Hexkit needs some changes introducing logic around event subscribers to correctly propagate tracing context across service boundaries as the autoinstrumentation for Kafka does not open a new child span, ending the propagation at the event subscriber.
+
```

Comment:
Is this not needed for synchronous requests?

Reply:
Not with autoinstrumentation enabled and does not need to be in hexkit.
Spans can be opened directly within in the FastAPI endpoint, but even that is not strictly necessary with autoinstrumentation enabled, span naming is just a bit more generic in that case.
If we do this manually, then we would want some middleware to handle context extraction and opening a parent span within the service.

I'll add a paragraph on that  to the docs.